### PR TITLE
Application gateway: Autoscale configuration

### LINKED ISSUE #2603: Support Application Gateway autoscaling
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.11.11
+ provider.azurerm v1.20.0

### Affected Resource(s)

* azurerm_application_gateway 

### Terraform Configuration Files

```hcl
variable "tenant_id" { }
variable "subscription_id" { }
variable "client_id" {}
variable "client_secret" {}
variable "location" {
    default = "centralus"
}

locals {
    ipAddressSku = "Standard"
    ipAddressAllocationType = "Static"
    resource_group_name = "terraform_appgw_test"
}

provider "azurerm" {
  tenant_id       = "${var.tenant_id}"
  subscription_id = "${var.subscription_id}"
  client_id       = "${var.client_id}"
  client_secret   = "${var.client_secret}"
}

resource "azurerm_resource_group" "test" {
  name     = "${local.resource_group_name}"
  location = "${var.location}"
}

resource "azurerm_virtual_network" "test" {
  name                = "virtualNetwork1"
  location            = "${azurerm_resource_group.test.location}"
  resource_group_name = "${azurerm_resource_group.test.name}"
  address_space       = ["10.0.0.0/16"]
}

resource "azurerm_subnet" "subnet1" {
  name                 = "subnet1"
  resource_group_name  = "${azurerm_resource_group.test.name}"
  virtual_network_name = "${azurerm_virtual_network.test.name}"
  address_prefix       = "10.0.1.0/24"
}

resource "azurerm_public_ip" "APPGW_01" {
  name = "APPGW-01"
  sku = "${local.ipAddressSku}"
  location = "${azurerm_resource_group.test.location}"
  resource_group_name = "${azurerm_resource_group.test.name}"
  public_ip_address_allocation = "${local.ipAddressAllocationType}"
}

locals {
  frontend_port_name_http = "http_frontend_port"
  frontend_port_name_https = "https_frontend_port"
  frontend_ip_configuration_name = "PublicFrontendIPConfig"
  appgw_backend_pool_name = "Untrusted"
}

resource "azurerm_application_gateway" "APPGW_01" {
  name = "APPGW-01"
  resource_group_name = "${azurerm_resource_group.test.name}"
  location = "${azurerm_resource_group.test.location}"

  sku {
    name = "Standard_v2"
    tier = "Standard_v2"
    capacity = "0" # I would expect 0 to work with the Standard_v2 sku, since 0 indicates auto-scaling.
  }

  gateway_ip_configuration {
    name = "gateway-ip-configuration"
    subnet_id = "${azurerm_subnet.subnet1.id}"
  }

  frontend_port {
    name = "${local.frontend_port_name_http}"
    port = 80
  }

  frontend_port {
    name = "${local.frontend_port_name_https}"
    port = 443
  }

  frontend_ip_configuration {
    name = "${local.frontend_ip_configuration_name}"
    public_ip_address_id = "${azurerm_public_ip.APPGW_01.id}"
  }

  backend_address_pool {
    name = "${local.appgw_backend_pool_name}"
  }

  backend_http_settings {
    name = "settings-http"
    cookie_based_affinity = "Disabled"
    port = 80
    protocol = "Http"
    request_timeout = 30
  }

  http_listener {
    name = "listener-http"
    frontend_ip_configuration_name = "${local.frontend_ip_configuration_name}"
    frontend_port_name = "${local.frontend_port_name_http}"
    protocol = "Http"
  }

  request_routing_rule {
    name = "rule-http"
    rule_type = "Basic"
    http_listener_name = "listener-http"
    backend_address_pool_name = "${local.appgw_backend_pool_name}"
    backend_http_settings_name = "settings-http"
  }
}
```

### Debug Output
https://gist.github.com/dmaltsiniotis/9d2259808586c081ea46c790999ba481

### Panic Output
N/A

### Expected Behavior

The application gateway validation should pass, and provision normally.

### Actual Behavior

The following error is observed when trying to run a plan:

Error: azurerm_application_gateway.APPGW_01: expected sku.0.capacity to be in the range (1 - 10), got 0

### Steps to Reproduce

1. `terraform plan'

### Important Factoids

Azure application gateway v2's are very new and behave differently from standard gateways. Even the azure web portal does not yet have all the configuration options for the v2 gateways. This _may_ necessitate a significant change to the validation logic for application gateways.

When setting the scaling to "automatic" in azure for a v2 application gateway, it's pulled back as having capacity zero by the azurerm provider for exiting resources.

### References

Here is the PR originally adding the SKU's, but not adjusting the capacity validation:
* #2015

### PR DESCRIPTION
**Pull request by timja**
Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/2603

### CODE REVIEW

File: azurerm/resource_arm_application_gateway.go [hunk 1]
```diff
@@ -2282,6 +2304,35 @@ func flattenApplicationGatewayRedirectConfigurations(input *[]network.Applicatio
...
+		configuration.MaxCapacity = utils.Int32(maxCapacity)
+	}
+
+	return &configuration
+}
+
+func flattenAutoscaleConfiguration(input *network.ApplicationGatewayAutoscaleConfiguration) []interface{} {
+	result := make(map[string]interface{})
+
+	result["min_capacity"] = *input.MinCapacity
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: We should nil check this. 
```suggestion
        if v := input.MinCapacity; v != nil {
          result["min_capacity"]= *v
        }
```

File: azurerm/resource_arm_application_gateway.go [hunk 2]
```diff
@@ -547,7 +547,23 @@ func resourceArmApplicationGateway() *schema.Resource {
...
 				},
 			},
-
+			"autoscale_configuration": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"min_capacity": {
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: We should add some validation for greater than zero and some max value. I can't find that information in the portal or the docs but testing for it should be fairly straightforward by just passing in a ridiculously large number. The API normally throws an error and lets the user know what the max value for that attribute is.

File: azurerm/resource_arm_application_gateway.go [hunk 3]
```diff
@@ -547,7 +547,23 @@ func resourceArmApplicationGateway() *schema.Resource {
...
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"max_capacity": {
+							Type:     schema.TypeInt,
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: We should add some validation for greater than zero and some max value. I can't find that information in the portal or the docs but testing for it should be fairly straightforward by just passing in a ridiculously large number. The API normally throws an error and lets the user know what the max value for that attribute is.

File: azurerm/resource_arm_application_gateway_test.go [hunk 4]
```diff
@@ -39,6 +39,63 @@ func TestAccAzureRMApplicationGateway_basic(t *testing.T) {
...
+				Config: testAccAzureRMApplicationGateway_autoscaleConfiguration(ri, testLocation()),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMApplicationGatewayExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "sku.0.name", "Standard_v2"),
+					resource.TestCheckResourceAttr(resourceName, "sku.0.tier", "Standard_v2"),
+					resource.TestCheckResourceAttr(resourceName, "autoscale_configuration.0.min_capacity", "2"),
+					resource.TestCheckResourceAttr(resourceName, "autoscale_configuration.0.max_capacity", "10"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.#", "0"),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway_test.go]: We should add an additional test step that updates these values as well.

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @timja. Thanks for submitting this PR! Things look great but we just need some additional validation/testing. Let me know if you have any questions.

File: azurerm/resource_arm_application_gateway.go (hunk 2 above)
Comment (timja) [reply azurerm/resource_arm_application_gateway.go]: Sure, can do. I didn't add it as I couldn't find a limit in docs the same as you, but will experiment when I can get back to this 👍

Comment (timja) [reply azurerm/resource_arm_application_gateway.go]: The portal allows between 2 and 10 for minimum, I've pushed a commit for this

File: azurerm/resource_arm_application_gateway.go (hunk 3 above)
Comment (timja) [reply azurerm/resource_arm_application_gateway.go]: The portal allows between `min_capacity` and 100 for this,
I've pushed a commit for this which does between 2 (min capacity minimum) and 100.
Unless there's an easy way to get the min capacity value (and its worth it), is this enough?

Comment (mbfrahry) [reply azurerm/resource_arm_application_gateway.go]: There is a very convoluted way of being able to get that minimum value but this is definitely sufficient. Thanks for taking the time to find this out

File: azurerm/resource_arm_application_gateway.go (hunk 2 above)
Comment (mbfrahry) [reply azurerm/resource_arm_application_gateway.go]: Great! That looks good

**Comment by timja** [comment]
All comments should be addressed

File: azurerm/resource_arm_application_gateway.go [hunk 5]
```diff
@@ -2282,6 +2306,37 @@ func flattenApplicationGatewayRedirectConfigurations(input *[]network.Applicatio
 	return results, nil
 }
 
+func expandAutoscaleConfiguration(d *schema.ResourceData) *network.ApplicationGatewayAutoscaleConfiguration {
+	vs := d.Get("autoscale_configuration").([]interface{})
+	v := vs[0].(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: This is panicking when `autoscale_configuration` hasn't been set. We can do something like the below to kick us out of this function if `autoscale_configuration` hasn't been set. 
```suggestion
      if len(vs) == 0 {
        return nil
      }
      v := vs[0].(map[string]interface{})
```

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @timja, apologies for missing this on my initial review but we hit quite a few panics after rerunning the application gateway tests. Do you mind taking a look at my comment and seeing if that or something similar makes sense to include?

```
=== RUN   TestAccAzureRMApplicationGateway_authCertificate
=== PAUSE TestAccAzureRMApplicationGateway_authCertificate
=== CONT  TestAccAzureRMApplicationGateway_authCertificate

------- Stderr: -------
panic: runtime error: index out of range

goroutine 400 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm.expandAutoscaleConfiguration(0xc0005f28c0, 0xc000a36c60)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_application_gateway.go:2311 +0x249
github.com/terraform-providers/terraform-provider-azurerm/azurerm.resourceArmApplicationGatewayCreateUpdate(0xc0005f28c0, 0x342c180, 0xc0005b6000, 0x2, 0x680a1a0)
```

File: azurerm/resource_arm_application_gateway.go [hunk 6]
```diff
@@ -2282,6 +2306,40 @@ func flattenApplicationGatewayRedirectConfigurations(input *[]network.Applicatio
...
+	if maxCapacity != 0 {
+		configuration.MaxCapacity = utils.Int32(maxCapacity)
+	}
+
+	return &configuration
+}
+
+func flattenAutoscaleConfiguration(input *network.ApplicationGatewayAutoscaleConfiguration) []interface{} {
+	result := make(map[string]interface{})
+
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]:
```
------- Stdout: -------
=== RUN   TestAccAzureRMApplicationGateway_zones
=== PAUSE TestAccAzureRMApplicationGateway_zones
=== CONT  TestAccAzureRMApplicationGateway_zones

------- Stderr: -------
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2715e1a]

goroutine 419 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm.flattenAutoscaleConfiguration(...)
	/opt/teamcity-agent/work/458e5e4800bd94f6/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_application_gateway.go:2333
github.com/terraform-providers/terraform-provider-azurerm/azurerm.resourceArmApplicationGatewayRead(0xc000554770, 0x342c180, 0xc001008000, 0xc00036cc20, 0x0)
```

Another panic cropped up.

```suggestion
    if input == nil {
        return []interface{}{result}
    }
```

**Review by mbfrahry — CHANGES_REQUESTED** [review]

File: azurerm/resource_arm_application_gateway.go [hunk 7]
```diff
@@ -2282,6 +2306,42 @@ func flattenApplicationGatewayRedirectConfigurations(input *[]network.Applicatio
 	return results, nil
 }
 
+func expandAutoscaleConfiguration(d *schema.ResourceData) *network.ApplicationGatewayAutoscaleConfiguration {
```
Comment (mcharriere) [inline azurerm/resource_arm_application_gateway.go]: Shouldn't this be called expandApplicationGatewayAutoscaleConfiguration for consistency?

File: azurerm/resource_arm_application_gateway.go [hunk 8]
```diff
@@ -2282,6 +2306,42 @@ func flattenApplicationGatewayRedirectConfigurations(input *[]network.Applicatio
...
+	}
+
+	if maxCapacity != 0 {
+		configuration.MaxCapacity = utils.Int32(maxCapacity)
+	}
+
+	return &configuration
+}
+
+func flattenAutoscaleConfiguration(input *network.ApplicationGatewayAutoscaleConfiguration) []interface{} {
```
Comment (mcharriere) [inline azurerm/resource_arm_application_gateway.go]: The same here. Isn't flattenApplicationGatewayAutoscaleConfiguration more appropriate?

**Review by mbfrahry — APPROVED** [review]
And all tests pass now! Thanks for your efforts here @timja. Merging now

**Pull request merged**